### DISCUSSION TITLE
Changing type of Context with Middleware

### DISCUSSION THREAD

**Discussion by grabbou** [trpc/trpc#2761] [Ideas]
Hey,

My context is:
```ts
type Context = { user: User | null }
```
User will be `null` when either token wasn't provided or provided token was invalid/expired. 

My router combines a few protected procedures (that require `user`) and a few that are public. 
```ts
const protectedProcedure = t.procedure.use(
  t.middleware(({ next, ctx }) => {
    if (!ctx.user) {
      throw new TRPCError({ code: "UNAUTHORIZED" });
    }
    return next({ ctx });
  })
);

const AppRouter = t.router({
   getDetails: protectedProcedure.query(({ ctx }) => { 
      console.log(ctx.user.id); // <--- TypeScript will complain that `user` can be null
   }),
   publicProcedure: t.procedure.query({})
})
```

Would it be possible for `t.middleware` to update the `Context` type based on what's passed to `next` as `ctx`? I know it's possible to perform this manually in every resolver, but middleware approach is just more elegant.

**Comment by KATT** [comment]
You have to give TypeScript a bit more help.

This will probably work:

```ts
const isAuthed = t.middleware(({ ctx, next }) => {
  if (!ctx.user) {
    throw new TRPCError({ code: 'UNAUTHORIZED' });
  }
  return next({
    ctx: {
      user: ctx.user, // user value is known to be non-null now
    },
  });
});

const protectedProcedure = t.procedure.use(isAuthed);
```

Note that you only need to give the _diff_ of new properties on the context when using a middleware

**Reply by KATT** [reply]
https://trpc.io/docs/v10/middlewares#context-swapping

**Reply by grabbou** [reply]
Oh, thanks for clarifying. I feel so bad now for missing this part of the documentation 🤦‍♂️ Thanks for swift reply!
